src/httpmessage.rs

```diff
@@ -260,6 +260,101 @@ pub trait HttpMessage {
         let boundary = Multipart::boundary(self.headers());
         Multipart::new(boundary, self)
     }
+    
+    /// Return stream of lines.
+    fn readlines(self) -> Readlines<Self>
+    where
+        Self: Stream<Item = Bytes, Error = PayloadError> + Sized,
+    {
+        Readlines::new(self)
+    }
+}
+
+/// Stream to read request line by line.
+pub struct Readlines<T>
+where
+    T: HttpMessage + Stream<Item = Bytes, Error = PayloadError> + 'static,
+{
+    req: T,
+    buff: Vec<u8>,
+    limit: usize,
+}
+
+impl<T> Readlines<T>
+where
+    T: HttpMessage + Stream<Item = Bytes, Error = PayloadError> + 'static,
+{
+    /// Create a new stream to read request line by line.
+    fn new(req: T) -> Self {
+        Readlines {
+            req,
+            buff: Vec::with_capacity(262_144),
+            limit: 262_144,
+        }
+    }
+    
+    /// Change max size of payload. By default max size is 256Kb
+    pub fn limit(mut self, limit: usize) -> Self {
+        self.limit = limit;
+        self
+    }
+}
+
+impl<T> Stream for Readlines<T>
+where
+    T: HttpMessage + Stream<Item = Bytes, Error = PayloadError> + 'static,
+{
+    type Item = String;
+    type Error = ReadlinesError;
+    
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        match self.req.poll() {
+            Ok(Async::Ready(Some(bytes))) => {
+                for b in bytes.iter() {
+                    if *b == '\n' as u8 {
+                        self.buff.push(*b);
```

Comment:
This is slooow 
Search for index then do extend_from_slice
Also you can try to use .split_to() and then extend

Reply:
Ok. fixing it

Reply:
what should I do if encoding isn't utf8 should I return Result::Err(EncodingError) from ::new()
and ::readlines() method

Reply:
```Rust
fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
    // check encoding
    if !self.checked_encoding {
        let encoding = self.req.encoding()?;
        let enc: *const Encoding = encoding as *const Encoding;
        if enc != UTF_8 {
            return Err(ReadlinesError::EncodingError)
        }
        self.checked_encoding = true;
    }
    
    // check if there is a newline in the buffer
    let mut found: Option<usize> = None;
    for (ind, b) in self.buff.iter().enumerate() {
        if *b == '\n' as u8 {
            found = Some(ind);
            break;
        }
    }
    if let Some(ind) = found {
        // check if line is longer than limit
        if ind+1 > self.limit {
            return Err(ReadlinesError::LimitOverflow);
        }
        let line = str::from_utf8(&self.buff.split_to(ind+1))?.to_owned();
        return Ok(Async::Ready(Some(line)));
    }
    // poll req for more bytes
    match self.req.poll() {
        Ok(Async::Ready(Some(mut bytes))) => {
            // check if there is a newline in bytes
            let mut found: Option<usize> = None;
            for (ind, b) in bytes.iter().enumerate() {
                if *b == '\n' as u8 {
                    found = Some(ind);
                    break;
                }
            }
            if let Some(ind) = found {
                // check if line is longer than limit
                if ind+1 > self.limit {
                    return Err(ReadlinesError::LimitOverflow);
                }
                let line = str::from_utf8(&bytes.split_to(ind+1))?.to_owned();
                // extend buffer with rest of the bytes;
                self.buff.extend_from_slice(&bytes);
                return Ok(Async::Ready(Some(line)));
            }
            self.buff.extend_from_slice(&bytes);
            Ok(Async::NotReady)
        },
        Ok(Async::NotReady) => Ok(Async::NotReady),
        Ok(Async::Ready(None)) => {
            if self.buff.len() == 0 {
                return Ok(Async::Ready(None));
            }
            if self.buff.len() > self.limit {
                return Err(ReadlinesError::LimitOverflow);
            }
            let line = str::from_utf8(&self.buff)?.to_owned();
            self.buff.clear();
            return Ok(Async::Ready(Some(line)))
        },
        Err(e) => Err(ReadlinesError::from(e)),
    }
}
```
is this good?

Reply:
You need to decide string

https://github.com/actix/actix-web/blob/master/src/extractor.rs#L439

Reply:
You should commit your code, it is hard comment on it

You should decide encoding only when you split new string 
So first poll, then search new line in chunk, then extend, then split, then decide encod Ng

Reply:
I search for a newline in the buffer first because there might be multiple newlines in chunk.
So I return a line if there is a newline in chunk. And extend the buffer with rest of it. Then I check
for newlines in the buffer new time. Is there a better way to do this.

Reply:
Ok, good.

Move ReadlinesError to errors module